opamp/observiq/reload_funcs.go

```diff
@@ -0,0 +1,262 @@
+
+		// Check if the updatable fields are equal
+		// If so then exit
+		if client.currentConfig.CmpUpdatableFields(newConfig) {
+			return false, nil
+		}
+
+		// Going to do an update prep a rollback
+		rollbackFunc, cleanupFunc, err := prepRollback(managerConfigPath)
+		if err != nil {
+			return false, fmt.Errorf("failed to prep for rollback: %w", err)
+		}
+
+		defer func() {
+			// Cleanup rollback
+			if err := cleanupFunc(); err != nil {
+				client.logger.Warn("Failed to cleanup rollback file", zap.Error(err))
+			}
+		}()
+
+		//create a copies for rollback
+		rollBackCfg := client.currentConfig.Copy()
+		rollbackIdent := client.ident.Copy()
+
+		// Updatable config fields
+		client.currentConfig.AgentName = newConfig.AgentName
+		client.currentConfig.Labels = newConfig.Labels
+
+		// Update identity
+		client.ident.agentName = newConfig.AgentName
+		client.ident.labels = newConfig.Labels
+
+		// Write out new config file
+		// Marshal back into bytes
+		newContents, err := yaml.Marshal(client.currentConfig)
+		if err != nil {
+			// Rollback file
+			if rollbackErr := rollbackFunc(); rollbackErr != nil {
+				client.logger.Error("Rollback failed for manager config", zap.Error(rollbackErr))
+			}
+			client.ident = rollbackIdent
+			client.currentConfig = *rollBackCfg
+			return false, fmt.Errorf("failed to reformat manager config: %w", err)
+		}
+
+		// Save config file to disk
+		if err := updateConfigFile(ManagerConfigName, managerConfigPath, newContents); err != nil {
+			// Rollback file
+			if rollbackErr := rollbackFunc(); rollbackErr != nil {
+				client.logger.Error("Rollback failed for collector config", zap.Error(rollbackErr))
+			}
+			client.ident = rollbackIdent
+			client.currentConfig = *rollBackCfg
+			return false, err
+		}
+
+		// Set the agent description
+		if err := client.opampClient.SetAgentDescription(client.ident.ToAgentDescription()); err != nil {
+			// Rollback file
+			if rollbackErr := rollbackFunc(); rollbackErr != nil {
+				client.logger.Error("Rollback failed for collector config", zap.Error(rollbackErr))
+			}
+			client.ident = rollbackIdent
+			client.currentConfig = *rollBackCfg
+			return false, fmt.Errorf("failed to set agent description: %w ", err)
+		}
+
+		return true, nil
+	}
+}
+
+func collectorReload(client *Client, collectorConfigPath string) opamp.ReloadFunc {
+	return func(contents []byte) (bool, error) {
+		rollbackFunc, cleanupFunc, err := prepRollback(collectorConfigPath)
+		if err != nil {
+			return false, fmt.Errorf("failed to prep for rollback: %w", err)
+		}
+
+		defer func() {
+			// Cleanup rollback
+			if err := cleanupFunc(); err != nil {
+				client.logger.Warn("Failed to cleanup rollback file", zap.Error(err))
+			}
+		}()
+
+		// Write new config file
+		if err := updateConfigFile(CollectorConfigName, collectorConfigPath, contents); err != nil {
+			return false, err
+		}
+
+		// Reload collector
+		if err := client.collector.Restart(context.Background()); err != nil {
```

Comment:
Do we have any concerns about losing the original context for the collector here? It seems like it works fine for now, but I'm wondering if it could cause problems down the line.

Reply:
There's a small concern. We call collector stop on a signal regardless. The collector just don't shut itself down on a signal now. I think it's fine. The alternative is to keep a reference to the signal context on the client and pass that through. I didn't like that but we could do it.

Reply:
Yeah, I think it's probably fine too, just oddly inconsistent between the first start of the collector and subsequent starts.

IMO, we should either do what you suggested (keep the ctx on the client and pass it in here), or pass in the background context on the initial startup; Either way, I think we should try to keep it consistent.

Reply:
Yeah I agree. I think we put the background context initially. That way it's clear the our program needs to shut down the collector.